external-crates/move/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -140,84 +126,728 @@ struct PackageWithResolverTOML<'a>(&'a Package);
+                                    *dep_name,
+                                    *is_other_external
+                                )?,
+                                other_pkg_deps
+                            ),
+                            format_deps(
+                                dep_path_from_root(root_package, &g, *dep_name, *is_external)?,
+                                pkg_deps
+                            ),
+                        )
+                    }
+                }
+            }
+
+            if internally_resolved {
+                // insert edges from the directly dependent package to its neighbors for
+                // internally resolved sub-graphs - due to how external graphs are constructed,
+                // edges between directly dependent packages and their neighbors are already in
+                // the sub-graph and would have been inserted in the first loop in this function
+                for to_pkg_name in g
+                    .package_graph
+                    .neighbors_directed(*dep_name, Direction::Outgoing)
+                {
+                    // unwrap is safe as all edges have a Dependency weight
+                    let sub_dep = g.package_graph.edge_weight(*dep_name, to_pkg_name).unwrap();
+                    self.package_graph
+                        .add_edge(*dep_name, to_pkg_name, sub_dep.clone());
+                }
+            }
+        }
+        Ok(())
     }
 
-    /// Create a dependency graph by reading a lock file.
+    /// Inserts a single direct dependency with given (package) name representing a sub-graph into
+    /// the combined graph. Returns true if the sub-graph has to be further merged into the combined
+    /// graph and false if it does not (i.e., if the dependency is already represented in the
+    /// combined graph).
+    fn insert_direct_dep(
+        &mut self,
+        dep: &PM::Dependency,
+        root_package: PM::PackageName,
+        dep_pkg_name: PM::PackageName,
+        sub_graph: &DependencyGraph,
+        mode: DependencyMode,
+        parent: &PM::DependencyKind,
+    ) -> Result<bool> {
+        match dep {
+            PM::Dependency::Internal(internal) => {
+                let PM::InternalDependency {
+                    kind,
+                    version,
+                    subst,
+                    digest,
+                    dep_override,
+                } = internal;
+
+                match self.package_table.get(&dep_pkg_name) {
+                    Some(_) => (),
+                    None => {
+                        let mut pkg = Package {
+                            kind: kind.clone(),
+                            version: version.clone(),
+                            resolver: None,
+                        };
+                        pkg.kind.reroot(parent)?;
+                        self.package_table.insert(dep_pkg_name, pkg);
+                    }
+                };
```

Comment:
```suggestion
                self.package_table
                    .entry(&dep_pkg_name)
                    .or_insert_with(|| {
                        let mut pkg = Package {
                            kind: kind.clone(),
                            version: version.clone(),
                            resolver: None,
                        };
                        pkg.kind.reroot(parent)?;
                        pkg
                    });
```

Reply:
The `reroot` function potentially erroring adds a wrinkle here...

Reply:
Ah yes, in that case:

```suggestion
                if let Entry::Vacant(entry) = self.package_table.entry(&dep_pkg_name) {
                    let mut pkg = Package {
                        kind: kind.clone(),
                        version: version.clone(),
                        resolver: None,
                    };
                    pkg.kind.reroot(parent)?;
                    entry.insert(dep_pkg_name, pkg);
                }
```